fs-cache/src/cache.rs

```diff
@@ -0,0 +1,668 @@
+use std::fs;
+use std::num::NonZeroUsize;
+use std::path::{Path, PathBuf};
+
+use data_error::{ArklibError, Result};
+use fs_atomic_light::temp_and_move;
+use fs_storage::utils::extract_key_from_file_path;
+use lru::LruCache;
+
+/// A cache entry that stores a value and its size in bytes.
+///
+/// This structure is used to track both the actual data (value)
+/// and its memory usage (size) in the cache.
+struct CacheEntry<V> {
+    value: V,
+    size: usize,
+}
+
+/// A combined in-memory and disk-based cache system.
+///
+/// This cache uses an LRU (Least Recently Used) eviction policy for the
+/// in-memory portion and persists data to disk for long-term storage.
+pub struct Cache<K, V> {
+    /// Label for logging
+    label: String,
+    /// Path to the underlying folder where data is persisted
+    path: PathBuf,
+    /// An in-memory LRU cache for quick access to frequently used items.
+    memory_cache: LruCache<K, CacheEntry<V>>,
+    /// The current memory usage in bytes.
+    current_memory_bytes: usize,
+    /// The maximum allowable memory usage in bytes.
+    max_memory_bytes: usize,
+    /// Whether to include values in debug logs
+    log_values: bool,
+}
+
+impl<K, V> Cache<K, V>
+where
+    K: Ord
+        + Clone
+        + std::fmt::Display
+        + std::hash::Hash
+        + std::str::FromStr
+        + std::fmt::Debug,
+    V: Clone + std::fmt::Debug + AsRef<[u8]> + From<Vec<u8>>,
+{
+    /// Creates a new cache instance.
+    ///
+    /// # Arguments
+    /// * `label` - Identifier used in logs
+    /// * `path` - Directory where cache files are stored
+    /// * `max_memory_bytes` - Maximum bytes to keep in memory
+    /// * `log_values` - Whether to include values in debug logs
+    pub fn new(
+        label: String,
+        path: &Path,
+        max_memory_bytes: usize,
+        log_values: bool,
+    ) -> Result<Self> {
+        let mut cache = Self {
+            label: label.clone(),
+            path: PathBuf::from(path),
+            // TODO: NEED FIX
+            memory_cache: LruCache::new(
+                NonZeroUsize::new(max_memory_bytes)
+                    .expect("Capacity can't be zero"),
+            ),
+            current_memory_bytes: 0,
+            max_memory_bytes,
+            log_values,
+        };
+
+        log::debug!(
+            "cache/{}: initialized with {} bytes limit",
+            label,
+            max_memory_bytes
+        );
+
+        cache.load_fs()?;
+        Ok(cache)
+    }
+
+    /// Retrieves a value by its key, checking memory first then disk.
+    /// Returns None if the key doesn't exist.
+    pub fn get(&mut self, key: &K) -> Option<V> {
+        log::debug!("cache/{}: retrieving value for key {}", self.label, key);
+
+        let value = self
+            .fetch_from_memory(key)
+            .or_else(|| self.fetch_from_disk(key));
```

Comment:
Here, I think this should be refactored to log whether the values were retrieved from memory or disk (if found).